### ISSUE TITLE
Evaluate if BSD Owl would be useful when rewriting the build system

### ISSUE THREAD

**Issue by michipili** [muellermichel/Hybrid-Fortran#40]
We had a little chat about that on hacknews… for almost a year! :)

  https://news.ycombinator.com/item?id=10144246

I am not sure if this is still actual, but if it is, why not starting to look at this?

**Comment by muellermichel** [comment]
Hi! Yea, completely forgot about that. It turned out I have a lot more other stuff to do until I can take care of the builds, but I will look into it when it comes around.

**Comment by michipili** [comment]
Maybe we can start listing the current problems with the actual build system, so that somebody else can also take a look into these things. :)

**Comment by muellermichel** [comment]
well the main limitation (it's not yet a *problem* since I don't need something more yet) is that the build system has two hardcoded target architectures: CPU and GPU. The hardcodedness leads to a lot of duplicated code in MakefileCommon and the inability to just compute the available architecture dynamically in a ./configure script. 

Now the default way to address this for me so far would be Makefile macros. I've already made some use of them. I'd just generate some files in configure for each available architecture and then parse the architecture names into a list in MakefileCommon and then loop over that list in macros for all the required target rules. 

Since this is close to a rewrite I'd be open for something else, however I'm very conservative about external dependencies. Basically it would have to be something that I can include in the repository or as a git submodule and then it needs to work even on top of quite old Red Hat linux (I can't even have Python 2.7 on there). Such is life on many HPC clusters.

**Comment by michipili** [comment]
> well the main limitation (it's not yet a problem since I don't need something more yet) is that the build system has two hardcoded target architectures: CPU and GPU.

Nice, this is precisely one of the use cases. :)  There is an example [in the source code](https://github.com/michipili/bsdowl/tree/master/example/langc/heat) the configurations are described in the `Configuration` directory. We say something like `make CONFIGURATION=Debug` or event set it from the environment.

See the [CI logs](https://travis-ci.org/michipili/bsdowl/jobs/135475576#L1043) if you want.

> Basically it would have to be something that I can include in the repository or as a git submodule and then it needs to work even on top of quite old Red Hat linux (I can't even have Python 2.7 on there). Such is life on many HPC clusters.

I think it is okay, the *bmake* program is really portable. I will try to install an old Red Hat Linux (which version?) on a VM to see what happens.  The only important dependency is GNU autoconf, but I do not expect it to be problematic to use an antiquated version.

> Oh and I should mention that it has to work without any sudo-privileges and without installing anything from the internet

That's fine, the CI already works without sudo. :)

**Comment by muellermichel** [comment]
Thank you for your interest and your input, @michipili! 

> That's fine, the CI already works without sudo. :)

How would you rollout Owl to a machine without internet access? Most people just shovel everything they need there with rsync - so the install process needs to be really straightforward. Bundling binaries creates all kinds of portability problems, so it's not ideal either. Basically if it could target GNU Make and only depend on scripts in already installed languages (Bash/Perl5/Python2.6) it would be the preferable way for HPC related applications since this is the most common way of doing it in this field.

**Comment by michipili** [comment]
Also, targetting GNU Make is no option, it is essentially incompatible with BSD Make. The procedures I use to install BSD Make and BSD Owl in the CI gives an idea on how to do this:

- https://github.com/michipili/anvil/blob/master/subr/anvil_travisci_autoinstall.sh#L21
- https://github.com/michipili/anvil/blob/master/subr/anvil_travisci_autoinstall.sh#L33

I am not sure to understand correctly what said about rsync. Am I right assuming you first prepare a “workspace” for your computation job using rsync and then submit a batch to the computation center. Can you cache some data (like the installed version of bmake?) or is everything cleaned up once your batch is done?

**Comment by muellermichel** [comment]
Ok, here is a typical setup of how people work in HPC:

[ Internet            ]
^
| git, http, etc.
|
v
[ work computer ]        --------- code: sftp/rsync --------------->  [ Big cluster, no internet, no sudo ]
                                    <------- results: sftp/rsync ---------------

In other words: Your stuff needs to be able to compile and run on the cluster *without* internet and *without* installing anything new to root. If you need a new binary on there with root access, in the best case you get it after a year (next update cycle). Cross-compiling for the cluster on the work computer is way too complicated for typical HPC users - they are scientists, not computer engineers. 

So, basically, I'd need a way of having Owl and all its dependencies archived so it can simply be loaded onto the cluster and then installed using a script running purely in userland. No curl, no git clone etc.

I actually think it's probably overkill. Probably it would be way easier to just amend my gmake based scheme to allow flexible architectures and pulling everything together in configure. I'd have to add a lot of stuff to Owl configure anyway (determining the exact GPU hardware/software setup on the target machine).
